hoomd/md/mesh/bond.py

```diff
@@ -0,0 +1,251 @@
+            self._mesh._add(self._simulation)
+        else:
+            if self._simulation != self._mesh._simulation:
+                raise RuntimeError("{} object's mesh is used in a "
+                                   "different simulation.".format(type(self)))
+
+        if not self.mesh._attached:
+            self.mesh._attach()
+
+        if isinstance(self._simulation.device, hoomd.device.CPU):
+            cpp_cls = getattr(_md, self._cpp_class_name)
+        else:
+            cpp_cls = getattr(_md, self._cpp_class_name + "GPU")
+
+        self._cpp_obj = cpp_cls(self._simulation.state._cpp_sys_def,
+                                self._mesh._cpp_obj)
+
+        self._apply_param_dict()
+        self._apply_typeparam_dict_with_mesh(self._cpp_obj, self.mesh)
+
+    def _apply_typeparam_dict_with_mesh(self, cpp_obj, mesh):
+        for typeparam in self._typeparam_dict.values():
+            try:
+                typeparam._attach_with_mesh(cpp_obj, mesh)
+            except ValueError as err:
+                raise err.__class__(
+                    f"For {type(self)} in TypeParameter {typeparam.name} "
+                    f"{str(err)}")
+
+    @property
+    def mesh(self):
+        """Mesh data structure used to compute the bond potential."""
+        return self._mesh
+
+    @mesh.setter
+    def mesh(self, value):
+        if self._attached:
+            raise RuntimeError("mesh cannot be set after scheduling.")
+        mesh = validate_mesh(value)
+        if self._added:
+            if mesh._added and self._simulation != mesh._simulation:
+                raise RuntimeError("Mesh and forces must belong to the same "
+                                   "simulation or SyncedList.")
+            self._mesh._add(self._simulation)
+        self._mesh = mesh
+
+    @property
+    def _children(self):
+        return [self.mesh]
+
+
+class Harmonic(MeshBond):
+    r"""Harmonic bond potential.
+
+    :py:class:`Harmonic` specifies a harmonic potential energy between all
+    particles that share an edge within the mesh. For more details see
+    :py:class:`hoomd.md.bond.Harmonic`.
+
+    Args:
+        mesh (:py:mod:`hoomd.mesh.Mesh`): Mesh data structure constraint.
+
+    Attributes:
+        params (TypeParameter[``mesh type``, dict]):
+            The parameter of the harmonic bonds for each particle type.
+            The dictionary has the following keys:
+
+            * ``k`` (`float`, **required**) - potential constant
+              :math:`[\mathrm{energy} \cdot \mathrm{length}^{-2}]`
+
+            * ``r0`` (`float`, **required**) - rest length
+              :math:`[\mathrm{length}]`
+
+    Examples::
+
+        harmonic = mesh.bond.Harmonic(mesh)
+        harmonic.params['wall'] = dict(k=3.0, r0=2.38)
+        harmonic.params['vesicle'] = dict(k=10.0, r0=1.0)
+    """
+    _cpp_class_name = "PotentialMeshBondHarmonic"
+
+    def __init__(self, mesh):
+        params = TypeParameter("params", "types",
```

Comment:
One way we could unify type specification would be to accept a type string or callable and the callable would return the valid types.